llarp/service/outbound_context.cpp

```diff
@@ -351,22 +350,36 @@ namespace llarp
       }
       // check for stale intros
       // update the introset if we think we need to
-      if (currentIntroSet.HasStaleIntros(now, path::intro_path_spread))
+      if (currentIntroSet.HasStaleIntros(now, path::intro_path_spread)
+          or remoteIntro.ExpiresSoon(now, path::intro_path_spread))
       {
         UpdateIntroSet();
+        ShiftIntroduction(false);
+      }
+
+      if (ReadyToSend())
+      {
+        if (not remoteIntro.router.IsZero() and not GetPathByRouter(remoteIntro.router))
+        {
+          // pick another good intro if we have no path on our current intro
+          std::vector<Introduction> otherIntros;
+          ForEachPath([now, router = remoteIntro.router, &otherIntros](auto path) {
+            if (path and path->IsReady() and path->Endpoint() != router
+                and not path->ExpiresSoon(now, path::intro_path_spread))
+            {
+              otherIntros.emplace_back(path->intro);
+            }
+          });
+          if (not otherIntros.empty())
+          {
+            std::shuffle(otherIntros.begin(), otherIntros.end(), CSRNG{});
+            remoteIntro = otherIntros[0];
+          }
+        }
       }
       // lookup router in intro if set and unknown
       if (not m_NextIntro.router.IsZero())
         m_Endpoint->EnsureRouterIsKnown(m_NextIntro.router);
-      // expire bad intros
-      auto itr = m_BadIntros.begin();
-      while (itr != m_BadIntros.end())
-      {
-        if (now > itr->second && now - itr->second > path::default_lifetime)
-          itr = m_BadIntros.erase(itr);
-        else
-          ++itr;
-      }
 
       if (ReadyToSend() and not m_ReadyHooks.empty())
       {
```

Comment:
`ReadyToSend()` checks for `GetPathByRouter()` success now, so the check after this is redundant.